crates/rpc/src/jsonrpc/websocket/logic.rs

```diff
@@ -0,0 +1,500 @@
+    use starknet_gateway_types::reply::Status;
+    use std::borrow::Cow;
+    use std::time::Duration;
+    use tokio::net::TcpStream;
+    use tokio::task::JoinHandle;
+    use tokio::time::timeout;
+    use tokio_tungstenite::tungstenite::Message;
+    use tokio_tungstenite::{connect_async, MaybeTlsStream, WebSocketStream};
+
+    #[tokio::test]
+    async fn params_are_required() {
+        let mut client = Client::new().await;
+
+        client
+            .send_request(&RpcRequest {
+                method: Cow::from(SUBSCRIBE_METHOD),
+                params: Default::default(),
+                id: RequestId::Null,
+            })
+            .await;
+
+        client
+            .expect_response(&RpcResponse {
+                output: Err(RpcError::InvalidParams),
+                id: RequestId::Null,
+            })
+            .await;
+
+        client.destroy().await;
+    }
+
+    #[tokio::test]
+    async fn can_subscribe() {
+        let mut client = Client::new().await;
+
+        let req_id = RequestId::Number(37);
+        client
+            .send_request(&RpcRequest {
+                method: Cow::from(SUBSCRIBE_METHOD),
+                params: RawParams(Some(&value(&Kind {
+                    kind: NEW_HEADS_TOPIC.into(),
+                }))),
+                id: req_id.clone(),
+            })
+            .await;
+
+        let expected_subscription_id = 0;
+        client
+            .expect_response(&successful_response(&expected_subscription_id, req_id).unwrap())
+            .await;
+
+        // Do this a bunch of times to ensure the test reception timeout is long enough.
+        for _i in 0..10 {
+            let header = header_sample();
+            client.head_sender.send_if_receiving(header.clone());
+
+            client
+                .expect_response(&SubscriptionItem {
+                    subscription_id: 0,
+                    item: header,
+                })
+                .await;
+        }
+
+        let req_id = RequestId::String("req_id".into());
+        client
+            .send_request(&RpcRequest {
+                method: Cow::from(UNSUBSCRIBE_METHOD),
+                params: RawParams(Some(&value(&SubscriptionId {
+                    id: expected_subscription_id,
+                }))),
+                id: req_id.clone(),
+            })
+            .await;
+        client
+            .expect_response(&successful_response(&true, req_id).unwrap())
+            .await;
+
+        // Now make sure we don't receive it. This is why testing the timeout was important.
+        client.head_sender.send_if_receiving(header_sample());
+        client.expect_no_response().await;
+
+        client.destroy().await;
+    }
+
+    // TODO Prevent duplicate subscriptions?
+    // This is actually tolerated by Alchemy, you can subscribe multiple times
+    // to the same topic and receive duplicated messages as a result.
+    // TODO Subscription limit?
```

Comment:
Yeah I thought about this; but only certain topics are properly duplicates. `newHeads` is ofcourse unique and silly for duplicates.

Probably we want to let the node owner configure a total subscription limit somehow. Some options:

1. event rate limit - this would be painful to implement, but is probably a more realistic cost estimation
2. limit active subscriptions per user

Probably though, the node operators that care about this will have their own layer on top to deal with these issues. Currently the only limiter is the buffer capacity which means the subscriber has to read fast enough..

Reply:
The [`governor` crate](https://docs.rs/governor/0.6.0/governor/) could help with that, but I'm not sure what the best behavior would be. And like you said, operators will add another, more effective layer with an authenticated logic on top, and probably segregate the nodes with websockets enabled.